sos_command.go

```diff
@@ -2,60 +2,113 @@ package main
 		Usage:  "Get metrics from object storage service",
+		Before: command.validate,
 		Action: command.execute,
 		Flags: []cli.Flag{
-			&cli.StringFlag{Name: "path-to-cluster-file", Aliases: []string{"clusters"}, Required: true, TakesFile: true,
-				Usage:       "File containing a list of clusters with the format server=token..",
-				Destination: &command.clusterFilePath,
+			&cli.StringFlag{
+				Name:        "k8s-server-url",
+				Aliases:     []string{"u"},
+				EnvVars:     []string{k8sServerUrlEnvVariable},
+				Required:    true,
+				Usage:       "A kubernetes server url from where to get the data from",
+				Destination: &command.clusterURL,
+			},
+			&cli.StringFlag{
+				Name:        "k8s-server-token",
+				Aliases:     []string{"t"},
+				EnvVars:     []string{k8sTokenEnvVariable},
+				Required:    true,
+				Usage:       "A kubernetes server token which has access to buckets",
+				Destination: &command.clusterToken,
+			},
+			&cli.StringFlag{
+				Name:        "database-url",
+				Aliases:     []string{"d"},
+				EnvVars:     []string{dbUrlEnvVariable},
+				Required:    true,
+				Usage:       "A postgres database url where to save relevant metrics",
+				Destination: &command.databaseURL,
+			},
+			&cli.StringFlag{
+				Name:        "exoscale-access-key",
+				Aliases:     []string{"k"},
+				EnvVars:     []string{keyEnvVariable},
+				Required:    true,
+				Usage:       "A key which has full access to an exoscale organization S3 storage",
+				Destination: &command.exoscaleKey,
+			},
+			&cli.StringFlag{
+				Name:        "exoscale-secret",
+				Aliases:     []string{"s"},
+				EnvVars:     []string{secretEnvVariable},
+				Required:    true,
+				Usage:       "A secret which has full access to an exoscale organization S3 storage",
+				Destination: &command.exoscaleSecret,
 			},
 		},
 	}
 }
 
 func (c *objectStorageCommand) execute(ctx *cli.Context) error {
 	log := AppLogger(ctx).WithName(objectStorageName)
-	ctrl.SetLogger(log)
-	log.Info("Loading Exoscale API credentials")
-	accessKey, secretKey, err := exoscale.LoadAPICredentials()
-	if err != nil {
-		return err
-	}
+
 	log.Info("Creating Exoscale client")
-	exoscaleClient, err := exoscale.CreateExoscaleClient(accessKey, secretKey)
+	exoscaleClient, err := exoscale.InitClient(c.exoscaleKey, c.exoscaleSecret)
 	if err != nil {
 		return err
 	}
-	log.Info("Reading input cluster file configuration")
-	clusters, err := kubernetes.ReadConf(c.clusterFilePath)
+
+	log.Info("Creating k8s client")
+	k8sClient, err := k8s.InitK8sClient(c.clusterURL, c.clusterToken)
 	if err != nil {
 		return err
 	}
-	log.Info("Creating k8s clients")
-	err = kubernetes.InitKubernetesClients(ctx.Context, clusters)
+
+	o := sos.NewObjectStorage(exoscaleClient, k8sClient, c.databaseInstance)
+	return o.Execute(ctx.Context)
+}
+
+func (c *objectStorageCommand) validate(ctx *cli.Context) error {
+	_ = LogMetadata(ctx)
+	log := AppLogger(ctx).WithName(objectStorageName)
+	ctrl.SetLogger(log)
```

Comment:
btw, with cli libary v2.17, you can add a flag-level action to validate a env var or cli flag: https://github.com/urfave/cli/pull/1337.

Reply:
Thanks for the input, update to v2.17.